feat(help):  `:helptags` in Lua (+Treesitter)

yochem opened the pull request:
Problem:
Tags are manually parsed in C which is not flexible and prone to errors. For example, switching the format of help files to e.g. markdown would require a large rewrite, while with this PR it would only require a TS query update. See #36249.

Solution:
Use the power of treesitter to extract the tags from helpfiles.

yochem commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,106 @@
+    if curtag == prevtag then
+      found = true
+      local other_fn = prevfn ~= curfn and (' and ' .. prevfn) or ''
+      vim.api.nvim_echo({
+        { ('E154: Duplicate tag "%s" in %s%s'):format(curtag, curfn, other_fn) }
+      }, true, { err = true })
```
Not sure if this is the right approach of reporting a Vim-style error

yochem commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,106 @@
+--- @param filename string helpfile with tags
+--- @return { [1]: string, [2]: string, [3]: string}[] tuple of tag, file, and search command
+local function extract_file_tags(filename)
+  local file = assert(io.open(filename, 'r'))
+  local source = file:read('*a')
+  file:close()
```
Is `assert()` fine here or should this report a Vim-style error to the user (E40)? Or ignore?

yochem commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,106 @@
+
+  if not check_duplicate_tags(tags) then
+    local tagsfile = vim.fs.joinpath(dir, 'tags')
+    local f = assert(io.open(tagsfile, 'w'))
```
See comment above: what to do on error?

clason commented:
Look at `src/nvim/lua/treesitter.c`.

yochem commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,178 @@
+  end)()
+end
+
+--- Returns a list of all existing "doc" subdirectories in the runtimepath.
```
Great suggestion! Didn't know about the function.

clason commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,164 @@
+    end)
+  end
+
+  coroutine.wrap(function()
```
I don't think we should add manual async at this point since we're already considering a `vim.async` module based on @lewis6991 work. Please look at #34009 and https://github.com/lewis6991/async.nvim, and pull in whatever you need from that into a `vim._async` module (that we'll merge with the one from #34009).

lewis6991 replied:
The potential of `vim.async` shouldn't restrict the use of coroutines. `vim.async` is just a very generalized layer on top of coroutines, but if the requirements are small (like) here, then I don't see a problem with using coroutines. I still think coroutines are important for generator functions (functions that return iterators).

The usage here is contained within ~30LOC, if anything this is might be preferable to using `vim.async`.

Though I'm not quite sure why coroutines are being used here (needs a comment). It looks like it's to not eat up the main loop, in which case `vim.async` + `uv.fs_open` might help in the future.

What is the effect if no coroutines where used here? Does creating tags from helpfiles cause lots of lag or something?

yochem replied:
I noticed it blocked the editor (~2 secs for the runtimedir). Although maybe vim.schedule will be enough for this. 

I added coroutines because I was looking for parallelization, as extracting tags for every file is a task that can be parallelized very efficiently. But I learned that coroutines all run on a single thread and file reads are blocking. Still, this implementation felt 'snappier' and faster to me than the fully sync approach. I'll try vim.schedule (+ uv.fs_open)

yochem commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,198 @@
+    async.await(vim.schedule)
+    vim.notify(('E153: Unable to open %s for reading'):format(filename), vim.log.levels.ERROR)
+    return tags
+  end
```
this is what the original code did. I can also just `assert` that the file is read if that's preferred.

justinmk commented on src/nvim/help.c:
```diff
@@ -1119,18 +833,12 @@ void ex_helptags(exarg_T *eap)
+  ADD_C(args, BOOLEAN_OBJ(add_help_tags));
+
+  Error err = ERROR_INIT;
+  NLUA_EXEC_STATIC("require('vim._helptags').generate(...)", args,
```
```suggestion
  NLUA_EXEC_STATIC("require('vim._core/help').gen_tags(...)", args,
```

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -624,6 +625,7 @@ add_custom_command(
       ${LUA_FILETYPE_MODULE_SOURCE} "vim.filetype"
       ${LUA_FS_MODULE_SOURCE} "vim.fs"
       ${LUA_F_MODULE_SOURCE} "vim.F"
+      ${LUA_HELPTAGS_MODULE_SOURCE} "vim._helptags"
```
not needed if it's moved to `_core` ... oh, i need to merge https://github.com/neovim/neovim/pull/35150 🙈

justinmk commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,198 @@
+  vim.validate('wait', wait, 'boolean', true)
+
+  local dirs = { vim.fs.normalize(dir) }
+  if dir == 'ALL' then
```
which way is more obvious and less surprising?

justinmk commented on runtime/lua/vim/_helptags.lua:
```diff
@@ -0,0 +1,198 @@
+    async.run(function()
+      async.join(tasks)
+    end):wait()
+  end
```
seems fine. are you trying to avoid the `wait` arg?

internal functions can be ugly :)

justinmk commented on runtime/lua/vim/_async.lua:
this can be dropped, right? this module is used by `vim.pack` already

justinmk replied:
> Or we could try to get the `vim.async` PR merged. As long as the core interface is stable

it is not ready to be merged. i raised a few issues in https://github.com/lewis6991/async.nvim and AFAICT there are still some pretty serious interface questions to resolve

> I'll see if I can rewrite it to work with the slimmer version

👍

justinmk commented on runtime/lua/vim/_core/help.lua:
```diff
@@ -0,0 +1,184 @@
+local M = {}
+
+local ts = vim.treesitter
+local async = vim._async
```
I'm not suggesting coroutines, I am suggesting simple synchronous code. If it takes 10s that is a regression vs the current C routine. Why does it take 10s?

yochem replied:
Is it mainly CI that you're concerned about execution speed? Because 10s without blocking the editor shouldn't be much of a problem for end users. But it's probably not ideal to slow every CI build with 10s... I'll probably close this PR then if I can't find a way to make it faster.

clason replied:
Funny how tree-sitter improvements leverage ;)

yochem commented on src/nvim/help.c:
```diff
@@ -1119,18 +835,14 @@ void ex_helptags(exarg_T *eap)
+
+  Error err = ERROR_INIT;
+  // NLUA_EXEC_STATIC("require('vim._core.help').gen_tags(...)", args,
+  //                  kRetNilBool, NULL, &err);
```
`module 'vim._core.help' not found`. Is this related to #36870?

justinmk replied:
If it is called with `nvim -ll` or `nvim -u NONE` then probably.

yochem committed:
feat(help): generate helptags using Treesitter

Problem:
Tags are manually parsed in C which is not flexible and prone to errors.
Extending the help system to allow for other formats (e.g. Markdown)
would require a large rewrite in the C core, while with Treesitter it
only needs a query update.

Solution:
Use the power of treesitter to extract the tags from helpfiles.

barrettruth commented:
Interested to hear y'alls thoughts on this. I sort of agree with justin on this but I thought of a different approach:

> Port all of the helptags C code to Lua; _and_ include this new implementation.

With this, we can always _try_ to use treesitter, and just fallback to the crappy scanner if the necessary runtime bits are not there.

Happy to help out however.

yochem commented:
That leaves the common case without treesitter: during build time TS is not available, so 99% of all helptags would be generated using the scanner. 

Forget the exact details too, but if `make doc` has access to the (then already built) Nvim binary, maybe we could build the `tags` file during that step?

barrettruth commented:
I'm a bit confused here. Why can't we just update [this](https://github.com/neovim/neovim/blob/6326ab3df9563d97c464b0a76b636b3ac98b4a70/runtime/CMakeLists.txt#L76-L77) to inject `VIMRUNTIME`?

I'd also like if this pr removed the associated C code it's replacing. Or are you planning to do this in a follow up?

Would be nice to move as much as possible of help stuff to lua - except of course any tag-related logic.

yochem committed:
refactor(help): call gen_tags via nlua_call_typval

nlua_call_typval reports Lua errors itself (E5108), so the manual
Error/emsg handling is no longer needed.

yochem committed:
fix(help): only accept tags delimited by whitespace

The old C parser only accepted a *tag* preceded by start-of-line or
whitespace and followed by whitespace or end-of-line. The vimdoc parser
also captures tags followed by other text, e.g.
"*$XDG_STATE_HOME*/.../logs" in starting.txt, which caused an E154
duplicate tag error for docs that were previously fine.

yochem committed:
build(runtime): set $VIMRUNTIME when generating helptags

The build-time nvim runs from the build dir and cannot find a runtime
directory relative to its executable, so vim.treesitter (needed by the
new helptags implementation) is not loadable. Set $VIMRUNTIME to the
source runtime for both helptags invocations, like cmake/Util.cmake
does for other generators.

yochem commented:
Thanks for the pointers! It seems to work indeed.

> I'd also like if this pr removed the associated C code it's replacing.

I think it does? See the [diff for help.c](https://github.com/neovim/neovim/pull/34277/changes#diff-4c61a6a0c127b3e3eda72ebe984f31fd395e203fac34f91a1873a73319054877)

BTW: think GH actions is down again, I'll rerun CI tomorrow. I'll also clean up commits and code a bit later.

barrettruth commented on runtime/lua/vim/_core/help.lua:
```diff
@@ -391,4 +393,152 @@ function M.local_additions()
+    error('duplicate tags')
+  end
+
+  -- vim.print('Helptags written to ' .. outpath)
```
remove

barrettruth commented on test/functional/ex_cmds/help_spec.lua:
```diff
@@ -316,3 +317,83 @@ describe(':help', function()
   end)
 end)
+
+describe(':helptags', function()
```
aren't there tests in `test_help.vim` that should now be removed/moved here? what do you think about moving them all to lua?

barrettruth commented on runtime/lua/vim/_core/help.lua:
```diff
@@ -391,4 +393,152 @@ function M.local_additions()
+    return a[1] < b[1]
+  end)
+
+  -- (3) check duplicates
```
This code is pretty much transpiled from the C function, right? I don't really like that the function returns one boolean and prints errors for all found duplicates, especially when the name is "find duplicates."

I think something like:

```lua
local duplicate_tags = find_duplicate_tags
vim.tbl_map(function(duplicate_tag) vim.api.nvim_err(...) end, duplicate_tags)
```

where duplicates is `string[]` makes more

non-blocker ofc. sorry, i know you've had enough back and forth in this pr....

barrettruth commented:
Thanks for getting right back on it - this is exciting!

barrettruth commented:
conflicts (exciting!)

justinmk commented on runtime/lua/vim/_core/help.lua:
```diff
@@ -395,4 +393,150 @@ function M.local_additions()
+  for _, tag in ipairs(tags) do
+    f:write(table.concat(tag, '\t') .. '\n')
+  end
+  f:close()
```
~~nit: can be simplified by using `vim.fn.writefile()`~~

eh, `writefile()` is a lateral move here bc we are writing the file as we go along.

justinmk commented on runtime/doc/deprecated.txt:
```diff
@@ -97,7 +97,7 @@ LUA
 UI
 
 • *hit-enter* *hit-enter-prompt*	With |ui2|, the legacy "Press ENTER"
-  *press-enter				prompt is never triggered.
+  *press-enter*				prompt is never triggered.
```
this was an invalid tag caught by the new checks.

justinmk commented on runtime/CMakeLists.txt:
```diff
@@ -41,12 +41,14 @@ foreach(PACKAGE ${PACKAGES})
-        -u NONE -i NONE -e --headless -c "helptags doc" -c quit
+      # "-c :helptags" (E154 duplicate tags, …) does not set the exit code, so check v:errmsg.
+      COMMAND ${CMAKE_COMMAND} -E env "VIMRUNTIME=${PROJECT_SOURCE_DIR}/runtime" ${NVIM_HOST_PRG}
+        -u NONE -i NONE -e --headless -c "helptags doc" -c "exe 'cquit' !empty(v:errmsg)"
```
fail this build step if `:helptags` finds errors / duplicate tags. this caught a problem on Windows (CRLF line endings), where we have been shipping a few invalid tags:

```
  D:/a/neovim/neovim/build/bin/nvim.exe -u NONE -i NONE -e --headless -c "helptags ++t doc" -c "exe 'cquit' !empty(v:errmsg)""                                                                                                                 
  Error in command line:                                                                                                                                                                                                                       
  E154: Duplicate tag "." in gui.txt and repeat.txt                                                                                                                                                                                            
  E154: Duplicate tag "/" in pattern.txt and usr_08.txt                                                                                                                                                                                        
  E154: Duplicate tag "/" in usr_08.txt and pattern.txt                                                                                                                                                                                        
  E154: Duplicate tag "/" in usr_08.txt                                                                                                                                                                                                        
  E154: Duplicate tag "/" in usr_08.txt  
```

this wasn't caught until now because the check in `gen_help_html.lua` parses via a buffer, and 'fileformat' strips CRLF before treesitter ever sees it.

This didn't happen with the C impl because it uses `os_fopen(…, "r")`  ( text mode) on Windows.

justinmk commented on runtime/CMakeLists.txt:
```diff
@@ -42,10 +42,11 @@ foreach(PACKAGE ${PACKAGES})
         ${PACKAGE} ${GENERATED_PACKAGE_DIR}/${PACKNAME}
       COMMAND ${NVIM_HOST_PRG}
-        -u NONE -i NONE -e --headless -c "helptags doc" -c quit
+        -u NONE -l ${PROJECT_SOURCE_DIR}/src/gen/gen_helptags.lua doc/tags doc
```
use the barebones helptag parser, like the zig build.

otherwise tree-sitter-vimdoc would be required to build, which is something we probably don't want to hassle with atm. (affects package maintainers/distros)

barrettruth commented:
thanks for pushing this over justin! glad we all figured this out.

yochem commented:
Yeah thanks Justin! Sorry I had so little time lately to finish it.
